subprojects/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/converters/IdeaModuleDependencyTargetNameMixin.java

```diff
@@ -29,7 +29,7 @@ public IdeaModuleDependencyTargetNameMixin(IdeaDependency ideaModuleDependency)
     public String getTargetModuleName() {
         if (ideaModuleDependency instanceof IdeaModuleDependency) {
             IdeaModuleDependency dependency = (IdeaModuleDependency) ideaModuleDependency;
-            return dependency.getDependencyModule().getName();
+            return dependency.getTargetModuleName();
```

Comment:
Reflection is exactly what we want, as this is the basic strategy the tapi uses for dealing with cross version changes. What we don't want to do, however, is some ad hoc specific for this specific property. We want to build some infrastructure that allows models to remove properties and this is the first time we've removed something.

We have 2 cases:
1. the model we receive from the target Gradle version provides a `targetModuleName` property, in which case we just want to use the value directly from the model (this happens reflectively btw)
2. the model we receive from the target Gradle version does not provide a `targetModuleName` but does provide a `dependencyModule` property whose value is a object with a `name` property. In this case we want to use the value of this `name` property.

The problem we have with the second case is that the logic that calculates the `targetModuleName` property in the second case expects that the model it receives implements the public interface that is _specific to the version of the Gradle that it is running in_, which won't have a `dependencyModule` property once we remove it. Instead, this logic should be presented with an interface that represents the model from the version of Gradle that is producing the model, that is an interface that has a `dependencyModule` property. The model doesn't actually have to implement this interface, it just needs to be structured so that it can be adapted to this interface, same as all the other cross-version stuff.

So, I would do something like this:

- Change `ProtocolToModelAdapter` so that it knows that it needs to adapt the constructor parameter provided to the mix in object from the actual model object to the type injected into the mix-in's constructor.
- Add a consumer-side interface with the old method on it.
- Inject this into the min-in's contructor instead of `IdeaDependency`.